include/boost/gil/image_processing/threshold.hpp

```diff
@@ -179,6 +181,131 @@ void threshold_truncate
     }
 }
 
+namespace detail{
+
+template <typename SrcView, typename DstView>
+void otsu_impl(SrcView const& src_view, DstView const& dst_view, threshold_direction direction)
+{
+    //deciding output channel type and creating functor
+    using source_channel_t = typename channel_type<SrcView>::type;
+    using result_channel_t = typename channel_type<DstView>::type;
+
+    std::array<std::size_t, 256> histogram{};
+    //initial value of min is set to maximum possible value to compare histogram data
+    //initial value of max is set to minimum possible value to compare histogram data
+    auto min = std::numeric_limits<source_channel_t>::max(),
+        max = std::numeric_limits<source_channel_t>::min();
+
+    if (sizeof(source_channel_t) > 1 || std::is_signed<source_channel_t>::value)
+    {
+        //iterate over the image to find the min and max pixel values
+        for (std::ptrdiff_t y = 0; y < src_view.height(); y++)
+        {
+            typename SrcView::x_iterator src_it = src_view.row_begin(y);
+            for (std::ptrdiff_t x = 0; x < src_view.width(); x++)
+            {
+                if (src_it[x] < min) min = src_it[x];
+                if (src_it[x] > min) min = src_it[x];
+            }
+        }
+
+        //making histogram
+        for (std::ptrdiff_t y = 0; y < src_view.height(); y++)
+        {
+            typename SrcView::x_iterator src_it = src_view.row_begin(y);
+
+            for (std::ptrdiff_t x = 0; x < src_view.width(); x++)
+            {
+                histogram[((src_it[x] - min) * 255) / (max - min)]++;
+            }
+        }
+    }
+    else
+    {
+        //making histogram
+        for (std::ptrdiff_t y = 0; y < src_view.height(); y++)
+        {
+            typename SrcView::x_iterator src_it = src_view.row_begin(y);
+
+            for (std::ptrdiff_t x = 0; x < src_view.width(); x++)
+            {
+                histogram[src_it[x]]++;
+            }
+        }
+    }
+
+    //histData = histogram data
+    //sum = total (background + foreground)
+    //sumB = sum background
+    //wB = weight background
+    //wf = weight foreground
+    //varMax = tracking the maximum known value of between class variance
+    //mB = mu background
+    //mF = mu foreground
+    //varBeetween = between class variance
+    //http://www.labbookpages.co.uk/software/imgProc/otsuThreshold.html
+    //https://www.ipol.im/pub/art/2016/158/
+    std::ptrdiff_t total_pixel = src_view.height() * src_view.width();
+    std::ptrdiff_t sum = 0, sumB = 0;
+    std::size_t wB = 0, wF = 0, threshold = 0;
+    double varMax = 0, mB, mF, varBetween;
```

Comment:
I'd really like to avoid the cryptic variable names and replace them with self-descriptive alternatives that are longer but also make the comments redundant. For example

```
sum_total
sum_back
weight_back
weight_fore
mean_back
mean_fore
var_intra_class
var_within_class
var_max
```

```